design/ceph/node-loss-rbd-rwo.md

```diff
@@ -0,0 +1,68 @@
+
+## Problem
+
+When a  node is lost where a pod is running with the RBD RWO volume is mounted. The volume cannot be mounted on another node since it could cause corruption if two clients are writing to the same volume.
+
+## Current Solution
+
+We have a manual solution to the problem which involves forceful deletion of a pod so that forced detachment and attachment work is possible. The problem with the current solution is that it takes around 11min to recover and also there are still chances of data corruption when the old pod on the lost node comes back online, this can cause multiple writers and lead to data corruption.
+
+## Proposed Solution
+
+> Note: This solution requires minimum kubernetes version 1.26.0
+
+We will use kubernetes beta feature [Non-Graceful Node Shutdown](https://kubernetes.io/blog/2022/12/16/kubernetes-1-26-non-graceful-node-shutdown-beta/) which will handle the volume detachment and attachment to the new node and deletion of the older pod on the lost node. To trigger this event for kubernetes we need to apply the taint  `out-of-service` manually.
+
+> Make sure to remove the taint `out-of-service` once the node is back in a good state.
+
+Once this taint is applied manually, rook will create [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md) using [csi-addons operator](https://github.com/csi-addons/kubernetes-csi-addons) which prevent ceph rbd client from communicating with lost node.
+
+After the new pod is running on the new node and the old node which was lost comes back, rook will delete the [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md).
+
+example of taint to be applied to lost node:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule
+```
+
+to remove the taint:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute-
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule-
+```
+
+Examples of NetworkFence CR:
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1alpha1
+kind: NetworkFence
+metadata:
+  name: <name> # We will keep the name the same as the node name
+  namespace: <ceph-cluster-namespace>
+spec:
+  driver: <driver-name> # for us it will `rook-ceph.rbd.csi.ceph.com`
+  fenceState: <fence-state> # For us it will be `Fenced`
+  cidrs:
+    - <nodeIP>
+  secret:
+    name: <csi-rbd-provisioner-secret-name>
+    namespace: <ceph-cluster-namespace>
+  parameters:
+    clusterID: <ceph-cluster-namespace>
+```
+
+> Note: This will be enabled by default in rook, in the case for some reason user wants to disable this feature in rook can edit `rook-ceph-operator-config` configmap and update the `ROOK_WATCH_FOR_NODE_FAILURE: "false"`.
+
```

Comment:
How about another section to discuss the possibility of adding the out-of-service taint automatically? I'm thinking we can provide a sample script that would add the taint automatically if a node is offline for too long (timeout would be configurable). The Rook operator wouldn't automate this, but the admin could choose to run the script if they want to automate it.

Reply:
added